### PR TITLE
o/h/ctlcmd: add a snapctl fail command

### PR DESCRIPTION
**Pull request by MiguelPires**
Add a `snapctl fail <reason>` command to rejects changes in a registry transaction in `change-view-<plug>` hooks.

### CODE REVIEW

**Review by zyga — COMMENTED** [review]
Should this be some sort of `snapctl debug fail <...>` command?

Separately I would make it take an argument that somehow implies where the failure is injected. As written it's somewhat generic in  naming but very specific in semantics which feels wrong.

**Comment by MiguelPires** [comment]
> Should this be some sort of `snapctl debug fail <...>` command?
>
> Separately I would make it take an argument that somehow implies where the failure is injected. As written it's somewhat generic in naming but very specific in semantics which feels wrong.

@zyga I don't think it should be a debug subcommand but maybe a registry subcommand?
I agree with the sentiment in general. At some point, I added a `--view` flag to relate this to registries, since we use it for other commands, but that also wasn't right. Maybe `snapctl registry fail <reason>` since I think we'll add `snapctl registry` sub-commands eventually. @pedronis wdyt?

File: overlord/hookstate/ctlcmd/fail_test.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+	s.state.Lock()
+	task := s.state.NewTask("run-hook", "my test task")
+	setup := &hookstate.HookSetup{Snap: "test-snap", Revision: snap.R(1), Hook: "other-hook"}
+	s.state.Unlock()
+
+	s.mockContext, err = hookstate.NewContext(task, s.state, setup, s.mockHandler, "")
+	c.Assert(err, IsNil)
+
+	stdout, stderr, err = ctlcmd.Run(s.mockContext, []string{"fail", "reason"}, 0)
+	c.Assert(err, ErrorMatches, i18n.G(`cannot use "snapctl fail" outside of a "change-view" hook`))
```
Comment (ZeyadYasser) [inline overlord/hookstate/ctlcmd/fail_test.go:114]: maybe we could also test for the ephemeral case?

**Review by ZeyadYasser — COMMENTED** [review]
The changes look really good, I agree regarding moving this as a registry sub-command `snapctl registry fail <reason>`. Also, It would be nice to have a small spread test with a simple test snap.

**Comment by MiguelPires** [comment]
Sorry, I thought I had commented here yesterday but looks like I forgot to click submit. Samuele's input was that this could be applicable to other hooks (so it would stay at `snapctl fail`) but to gate this behind the feature flag for now until we can confirm that

File: overlord/hookstate/ctlcmd/fail_test.go (hunk 1 above)
Comment (MiguelPires) [reply overlord/hookstate/ctlcmd/fail_test.go:125]: Added, thanks

**Review by ZeyadYasser — APPROVED** [review]
LGTM

**Review by andrewphelpsj — APPROVED** [review]
Thanks!

**Pull request merged**